File: examples/java/kora-java-petclinic/src/main/java/io/koraframework/example/petclinic/repository/OwnerRepository.java
package io.koraframework.example.petclinic.repository;

import io.koraframework.database.common.UpdateCount;
import io.koraframework.database.common.annotation.Id;
import io.koraframework.database.common.annotation.Query;
import io.koraframework.database.common.annotation.Repository;
import io.koraframework.database.jdbc.JdbcRepository;
import io.koraframework.example.petclinic.model.dao.Owner;
import java.util.List;
import java.util.Optional;

@Repository
public interface OwnerRepository extends JdbcRepository {

    @Query("SELECT %{return#selects} FROM %{return#table} WHERE id = :id")
    Optional<Owner> findById(long id);

    @Query("""
            SELECT %{return#selects}
            FROM %{return#table}
            WHERE lower(last_name) LIKE lower(:lastNamePattern)
            ORDER BY last_name, first_name
            """)
    List<Owner> findByLastName(String lastNamePattern);

    @Id
    @Query("INSERT INTO %{entity#inserts -= id}")
    long insert(Owner entity);

    @Query("UPDATE %{entity#table} SET %{entity#updates} WHERE %{entity#where = @id}")
    UpdateCount update(Owner entity);
}

File: examples/kotlin/kora-kotlin-petclinic/src/main/kotlin/io/koraframework/kotlin/example/petclinic/repository/OwnerRepository.kt
package io.koraframework.kotlin.example.petclinic.repository

import io.koraframework.database.common.UpdateCount
import io.koraframework.database.common.annotation.Id
import io.koraframework.database.common.annotation.Query
import io.koraframework.database.common.annotation.Repository
import io.koraframework.database.jdbc.JdbcRepository
import io.koraframework.kotlin.example.petclinic.model.Owner

@Repository
interface OwnerRepository : JdbcRepository {
    @Query("SELECT %{return#selects} FROM %{return#table} WHERE id = :id")
    fun findById(id: Long): Owner?

    @Query("""
        SELECT %{return#selects}
        FROM %{return#table}
        WHERE lower(last_name) LIKE lower(:lastNamePattern)
        ORDER BY last_name, first_name
    """)
    fun findByLastName(lastNamePattern: String): List<Owner>

    @Id
    @Query("INSERT INTO %{entity#inserts -= id}")
    fun insert(entity: Owner): Long

    @Query("UPDATE %{entity#table} SET %{entity#updates} WHERE %{entity#where = @id}")
    fun update(entity: Owner): UpdateCount
}
